pallets/avn-transaction-payment/src/tests/adjustment_fee_tests.rs

```diff
@@ -0,0 +1,582 @@
+
+                    assert!(event_emitted(&mock::RuntimeEvent::AvnTransactionPayment(
+                        crate::Event::<TestRuntime>::AdjustedTransactionFeePaid {
+                            who: sender,
+                            fee: percentage_value as u128
+                        }
+                    )));
+                })
+            }
+
+            #[test]
+            fn and_percentage_fee_is_100() {
+                new_test_ext().execute_with(|| {
+                    let sender = to_acc_id(1u64);
+                    set_initial_sender_balance(&sender);
+
+                    let high_percentage_fee = 100;
+                    let config = AdjustmentInput::<TestRuntime> {
+                        fee_type: FeeType::PercentageFee(PercentageFeeConfig {
+                            percentage: high_percentage_fee,
+                            _marker: sp_std::marker::PhantomData::<TestRuntime>,
+                        }),
+                        adjustment_type: AdjustmentType::None,
+                    };
+
+                    assert_ok!(AvnTransactionPayment::set_known_sender(
+                        RuntimeOrigin::root(),
+                        sender,
+                        config,
+                    ));
+                    assert_eq!(AvnTransactionPayment::is_known_sender(sender), true);
+
+                    pay_gas_and_call_remark(&sender);
+                    assert_eq!(INITIAL_SENDER_BALANCE, Balances::free_balance(sender));
+
+                    assert!(event_emitted(&mock::RuntimeEvent::AvnTransactionPayment(
+                        crate::Event::<TestRuntime>::AdjustedTransactionFeePaid {
+                            who: sender,
+                            fee: 0u128
+                        }
+                    )));
+                })
+            }
+        }
+
+        mod call_is_set_with_valid_transaction_based_adjustment_type {
+            use super::*;
+            use crate::mock::AvnTransactionPayment;
+
+            #[test]
+            fn and_valid_fixed_fee() {
+                new_test_ext().execute_with(|| {
+                    let sender = to_acc_id(1u64);
+                    set_initial_sender_balance(&sender);
+
+                    let number_of_transactions = 1;
+                    let config = AdjustmentInput::<TestRuntime> {
+                        fee_type: FeeType::FixedFee(FixedFeeConfig { fee: FIXED_FEE }),
+                        adjustment_type: AdjustmentType::TransactionBased(NumberOfTransactions {
+                            number_of_transactions,
+                        }),
+                    };
+
+                    assert_ok!(AvnTransactionPayment::set_known_sender(
+                        RuntimeOrigin::root(),
+                        sender,
+                        config,
+                    ));
+                    assert_eq!(AvnTransactionPayment::is_known_sender(sender), true);
+
+                    pay_gas_and_call_remark(&sender);
```

Comment:
It would be so much easier to review the tests if all of this were captured into a single function called `check_sender_paid_fee(sender, fee)` making sure that the event and the balance change are in sync.

Also, removing the difference in abstraction level. When reading the code at this level, I can see:
- prepare balance
- make payment
- check fees
- prepare balance
- make payment
- check different fees (if the fee conditions have expired, for example)

This is not a request for change, now, but just an awareness example of how it could save time in reviewing (and in debugging tests six months down the line) with just a small approach to how to write tests.

Reply:
I agree, the tests have been refactored with the `check_sender_paid_fee` function and some more function to make it easier to read.